Add `align_offset` intrinsic and `[T]::align_to` function

oli-obk opened the pull request:
[Rendered](https://github.com/oli-obk/rfcs/blob/is_aligned/text/0000-is-aligned-intrinsic.md)

cc @dwrensha @eddyb @joshlf @nikomatsakis 

Background: https://github.com/solson/miri/issues/190 and https://github.com/rust-lang/rust/issues/42561

Screwtapello commented:
Under "Motivation":

> The standard library (and most likely many crates) [...] check whether a pointer is aligned in order to perform optimizations like reading multiple bytes at once. Not only is this code which is easy to get wrong, and which is hard to read (and thus increasing the chance of future breakage) but it also makes it impossible for miri to evaluate such statements. This means that miri cannot do utf8-checking, since that code contains such optimizations.

I assume it does something like go byte-by-byte up until the alignment of a `u64`, then the SIMD accelleration kicks in?

But later, under "How We Teach This":

> This means that while !is_aligned(ptr) { ... } should be considered a potential infinite loop.

...so this intrinsic could not be relied on for things like the UTF-8 validation mentioned in the motivation section?

scottmcm commented:
I'm not certain that this is the right public API.  For example, I was looking at the [mem::swap loop](https://github.com/rust-lang/rust/blob/4711982314b/src/libcore/mem.rs#L528) recently, and wondering whether it'd be worth aligning one of the three copies.  But the way I'd default to writing that would be to mod by alignment to memcpy the prefix first, not a loop `while !is_aligned`.  It also feels a bit odd that this requires creating an improperly-aligned pointer first, then checking if it was ok.  Maybe there's a way that this could instead produce the properly-aligned pointer from the uncertain one?  Or maybe even the desired API is more like `&[T]` -> `(&[T], &[U], &[T])`, to explicitly separate off the unaligned prefix and suffix...

And a minor thing: the title of this makes me think, "intrinsics are internal implementation details; do they even need RFCs?".  I'd like the "this is proposed for stability in core&std" to be more visible, such as at least in the summary (not just a line at the bottom of the design).

nagisa commented:
Okay so I find it kinda funny that MIRI is not expected to support checking
alignment the old way (producing the dummy integer representations of
integers). Sure we'll add the intrinsic but what about

1. Code that checks for alignment that's not natural to the type? (I do not
remember there being an argument for the size of alignment and can't check
ATM);
2. Something like corrode converted code/code written by a C programmer who
would do the test the old way;

Then?

Furthermore it seems to me that MIRI is in a pecular spot where it has to
just support unaligned access for all types. So that even if it's (host's)
alignment requirements are different from the target requirements it has to
just work anyway.

On Jun 28, 2017 05:00, "Screwtapello" <notifications@github.com> wrote:

> Under "Motivation":
>
> The standard library (and most likely many crates) [...] check whether a
> pointer is aligned in order to perform optimizations like reading multiple
> bytes at once. Not only is this code which is easy to get wrong, and which
> is hard to read (and thus increasing the chance of future breakage) but it
> also makes it impossible for miri to evaluate such statements. This means
> that miri cannot do utf8-checking, since that code contains such
> optimizations.
>
> I assume it does something like go byte-by-byte up until the alignment of
> a u64, then the SIMD accelleration kicks in?
>
> But later, under "How We Teach This":
>
> This means that while !is_aligned(ptr) { ... } should be considered a
> potential infinite loop.
>
> ...so this intrinsic could not be relied on for things like the UTF-8
> validation mentioned in the motivation section?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/2043#issuecomment-311535521>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AApc0piiCPioXnqDNLBizijqrN__IMfDks5sIbO2gaJpZM4OFPYw>
> .
>

nagisa commented:
Also  i remember using this function from bluss crates which looks like
this:

fn alignto<T>(&[u8]) -> (&[u8], &[T], &[u8])

Perhaps we could just add that to the standard library and make Miri
somehow recognise it?

On Jun 28, 2017 7:56 AM, "Simonas Kazlauskas" <s@kazlauskas.me> wrote:

> Okay so I find it kinda funny that MIRI is not expected to support
> checking alignment the old way (producing the dummy integer representations
> of integers). Sure we'll add the intrinsic but what about
>
> 1. Code that checks for alignment that's not natural to the type? (I do
> not remember there being an argument for the size of alignment and can't
> check ATM);
> 2. Something like corrode converted code/code written by a C programmer
> who would do the test the old way;
>
> Then?
>
> Furthermore it seems to me that MIRI is in a pecular spot where it has to
> just support unaligned access for all types. So that even if it's (host's)
> alignment requirements are different from the target requirements it has to
> just work anyway.
>
> On Jun 28, 2017 05:00, "Screwtapello" <notifications@github.com> wrote:
>
>> Under "Motivation":
>>
>> The standard library (and most likely many crates) [...] check whether a
>> pointer is aligned in order to perform optimizations like reading multiple
>> bytes at once. Not only is this code which is easy to get wrong, and which
>> is hard to read (and thus increasing the chance of future breakage) but it
>> also makes it impossible for miri to evaluate such statements. This means
>> that miri cannot do utf8-checking, since that code contains such
>> optimizations.
>>
>> I assume it does something like go byte-by-byte up until the alignment of
>> a u64, then the SIMD accelleration kicks in?
>>
>> But later, under "How We Teach This":
>>
>> This means that while !is_aligned(ptr) { ... } should be considered a
>> potential infinite loop.
>>
>> ...so this intrinsic could not be relied on for things like the UTF-8
>> validation mentioned in the motivation section?
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rust-lang/rfcs/pull/2043#issuecomment-311535521>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AApc0piiCPioXnqDNLBizijqrN__IMfDks5sIbO2gaJpZM4OFPYw>
>> .
>>
>

oli-obk commented:
> 1. Code that checks for alignment that's not natural to the type? (I do not remember there being an argument for the size of alignment and can't check ATM);

You can always cast your pointer to a type supporting your desired alignment.

> 2. Something like corrode converted code/code written by a C programmer who would do the test the old way;

Well... right now we already don't support pointer arithmetic inside constants, so adding this feature will actually allow more code to work in const eval. You can always write code that won't be part of const eval.

> Okay so I find it kinda funny that MIRI is not expected to support checking alignment the old way 

This feature is needed for constant evaluation. CTFE cannot ever treat a pointer into any position of  `[u8; N]` as aligned to an alignment other than `1`, because it could either be `&arr[0]` which is aligned, or `&arr[1]` or... you don't know before you run the program and the array is placed at some arbitrary memory location. I mean we could decide that any const eval type is aligned like `usize`, which is what C does (`malloc` guarantees that you get at least the same alignment that the pointer type has). Rust doesn't guarantee this, even if jemalloc does it (tested by allocating a few `u8`s on the heap).

But even if we guarantee pointer alignment, we'd have the same issue for alignments above pointer alignment that we have right now for alignments bigger than `1`.

> ...so this intrinsic could not be relied on for things like the UTF-8 validation mentioned in the motivation section?

Well... the utf8-validation does not rely on the optimization to ever trigger. It would just keep on doing the unoptimized stuff.

>  fn alignto<T>(&[u8]) -> (&[u8], &[T], &[u8])
> the desired API is more like &[T] -> (&[T], &[U], &[T]),

That would work, as long as I'm allowed to implement it as

```rust
fn alignto<T>(slice: &[u8]) -> (&[u8], &[T], &[u8]) {
    (slice, &[], &[])
}
```

Otherwise we have gained nothing. This API has the advantage over `is_aligned` of being hard to use wrongly.

nagisa commented:
On Jun 28, 2017 10:14, "Oliver Schneider" <notifications@github.com> wrote:


   1. Code that checks for alignment that's not natural to the type? (I do
   not remember there being an argument for the size of alignment and can't
   check ATM);

You can always cast your pointer to a type supporting your desired
alignment.


This is not true. There's no type with natural alignment of 4KB which you
might end up doing to align data to a page for example.


   1. Something like corrode converted code/code written by a C programmer
   who would do the test the old way;

Well... right now we already don't support pointer arithmetic inside
constants, so adding this feature will actually allow more code to work in
const eval. You can always write code that won't be part of const eval.

Okay so I find it kinda funny that MIRI is not expected to support checking
alignment the old way

This feature is needed for constant evaluation. CTFE cannot ever treat a
pointer into any position of [u8; N] as aligned to an alignment other than 1,
because it could either be &arr[0] which is aligned, or &arr[1] or... you
don't know before you run the program and the array is placed at some
arbitrary memory location. I mean we could decide that any const eval type
is aligned like usize, which is what C does (malloc guarantees that you get
at least the same alignment that the pointer type has). Rust doesn't
guarantee this, even if jemalloc does it (tested by allocating a few u8s on
the heap).

But even if we guarantee pointer alignment, we'd have the same issue for
alignments above pointer alignment that we have right now for alignments
bigger than 1.

...so this intrinsic could not be relied on for things like the UTF-8
validation mentioned in the motivation section?

Well... the utf8-validation does not rely on the optimization to ever
trigger. It would just keep on doing the unoptimized stuff.

fn alignto(&[u8]) -> (&[u8], &[T], &[u8])

the desired API is more like &[T] -> (&[T], &[U], &[T]),

That would work, as long as I'm allowed to implement it as

fn alignto<T>(slice: &[u8]) -> (&[u8], &[T], &[u8]) {
    (slice, &[], &[])
}

Otherwise we have gained nothing. This API has the advantage over is_aligned
of being hard to use wrongly.


Yes that would be a valid implementation for the purposes of CTFE. That is
I believe that CTFE not handling alignment in any way is beneficial, even,
as it is a detail a number of layers below it. I do not exactly see an use
case for ctfe-time alignment handling either.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/rust-lang/rfcs/pull/2043#issuecomment-311576970>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AApc0oOJiw-xqWH4sGa9-wXE9wLfkIy5ks5sIf07gaJpZM4OFPYw>
.

oli-obk commented:
Ok, I agree that `is_aligned` has too many problems to be useful, and it's not the most elegant solution to the actual problem we're trying to solve. I'll change the rfc to suggest an intrinsic for splitting a slice into aligned and nonaligned parts

@nagisa the function splitting slices also doesn't support your 4KB alignment use case. But it would be addressed once `#[repr(align(4096))]` (https://github.com/rust-lang/rust/issues/33626) is through, since you can just create a wrapper type with the correct alignment.

ubsan commented:
@nagisa "Furthermore it seems to me that MIRI is in a pecular spot where it has to
just support unaligned access for all types" I don't get why this is? I feel like, on UB, miri should just error, not continue working.

scottmcm commented:
Thanks for the update, @oli-obk!  Sorry for being slow to respond.

A few thoughts:

- I think this could benefit from the RFC https://github.com/rust-lang/rfcs/pull/2059 split.  The reference/implementation information here is good, but I think it could use a bit more of "here's how you use this".  Maybe show how the examples in the motivation change?
- Is it possible that the intrinsic and what's exposed should be different?  As a strawman, it might be fine for the intrinsic to be `fn unaligned_prefix_len<T>(p: *const (), n: usize) -> usize`, with the method in core for eventual stabilization providing the nice slice-oriented interface atop it.  Or the intrinsic could take the desired alignment as a parameter even if the core function uses type alignment, etc.
- Why always `u8`?  It seems reasonable that, say, a `sum` function would want to use `fn (&[f32]) -> (&[f32], &[simd::f32x8], &[f32])`.
- nitpick: You mention the function is unsafe, so the code blocks and signatures should probably have `unsafe` in them.

oli-obk commented:
> I think this could benefit from the RFC #2059 split. The reference/implementation information here is good, but I think it could use a bit more of "here's how you use this". Maybe show how the examples in the motivation change?

Done. Plus added a note that we should extend the `align` documentation to mention the `alignto` functions and intrinsics

> Is it possible that the intrinsic and what's exposed should be different? As a strawman, it might be fine for the intrinsic to be fn unaligned_prefix_len<T>(p: *const (), n: usize) -> usize, with the method in core for eventual stabilization providing the nice slice-oriented interface atop it. Or the intrinsic could take the desired alignment as a parameter even if the core function uses type alignment, etc.

Not just possible, but necessary. The optimizations in the stdlib can be very frickly (you can't always use iterators or loops). I added the most basic intrinsic that covers all use cases (basically the intrinsic is a more advanced form of the originally suggested `is_aligned` intrinsic), but kept the library functions.

> Why always u8? It seems reasonable that, say, a sum function would want to use fn (&[f32]) -> (&[f32], &[simd::f32x8], &[f32]).

Just an oversight. Fixed.

> nitpick: You mention the function is unsafe, so the code blocks and signatures should probably have unsafe in them.

Done.

scottmcm commented on text/0000-is-aligned-intrinsic.md:
```diff
@@ -0,0 +1,287 @@
+
+Add an intrinsic (`fn alignto(ptr: *const (), align: usize) -> usize`)
+which returns the number of bytes that need to be skipped in order to correctly align the
+pointer `ptr` to `align`.
```
bikeshed: now that this doesn't return a pointer, maybe it should have a different name?  Would help make it easier to talk about the two things, too, if the intrinsic and the eventually-stable methods had different names.

oli-obk replied:
Makes sense. I chose `align_offset` for now, although `byte_offset_to_aligned` might be clearer if wordier.

scottmcm commented on text/0000-is-aligned-intrinsic.md:
```diff
@@ -0,0 +1,287 @@
+## standard library functions
+
+Add a new method `alignto` to `*const T` and `*mut T`, which forwards to the
+`alignto`
```
I'm not sure I like this, since the intrinsic signature is designed to be easy and efficient for a backend to implement, at the cost of harder to call.  Is the slice version insufficient?

oli-obk replied:
Yes, the slice version creates all three slices, even in the cases where you might notice that you don't need the second and third slice while processing the first one. This is a real case  as shown  in an example further down.

scottmcm commented on text/0000-is-aligned-intrinsic.md:
```diff
@@ -0,0 +1,287 @@
+        let tail = &tail[mid_count * core::mem::size_of::<T>()..];
+        (head, mid, tail)
+    } else {
+        // can't properly fit a T into a sequence of `U`
```
Maybe just `panic!`?  Or static_assert once https://github.com/rust-lang/rust/pull/42859 makes size_of `const`?

oli-obk replied:
We don't have `static_assert` :(

oli-obk replied:
We could simply lint against it in the non-monomorphization cases

scottmcm commented on text/0000-is-aligned-intrinsic.md:
```diff
@@ -0,0 +1,287 @@
+        // due to `size_of::<T>() % size_of::<U>() == 0`,
+        // the fact that `size_of::<U>() > align_of::<U>()`,
+        // and the fact that `align_of::<T>() > align_of::<U>()` if `offset != 0` we know
+        // that `offset % source_size == 0`
```
nit: technically we don't know this if the intrinsic returns `usize::max()`, right?

oli-obk replied:
True, but in that case we get a value way bigger than `slice.len()` and thus result in the case where `head == slice` and `mid` and `tail` are empty. Which is the correct behaviour.

```rust
let head_count = offset / source_size;
let split_position = core::cmp::max(slice.len(), head_count);
let (head, tail) = slice.split_at(split_position);
```

But I just noticed that for zsts this function panics due to division by zero. Not sure what to do about that.

scottmcm replied:
Maybe just a ZST check at the beginning that just returns `(slice, &[], &[])`?  It's not like aligned reads of nothing are better than unaligned reads of nothing :laughing:

scottmcm commented on text/0000-is-aligned-intrinsic.md:
```diff
@@ -0,0 +1,287 @@
+on all current platforms. `alignto_mut` is expanded accordingly.
+
+Any backend may choose to always produce the following implementation, no matter
+the given arguments, since the intrinsic is purely an optimization aid.
```
Is this text is left-over from when the slice version was the intrinsic?  Seems like now it'll be something more like "if the backend compiles the intrinsic as `return usize::max_value()`, this method will optimize down to the following"

scottmcm commented on text/0000-is-aligned-intrinsic.md:
````diff
@@ -0,0 +1,287 @@
+
+```rust
+let ptr = v.as_ptr();
+let align = ptr.offset(index).alignto(usize_bytes);
````
It's surprising to me that the text above says "Once can clearly see the three slice parts", but this code doesn't use the slice method.  Also, consider linking to the full code -- I think, from the other issue, that it's https://github.com/rust-lang/rust/blob/f09576c4a41727a8d10bbfd8fd3fb2e10e1be3b3/src/libcore/str/mod.rs#L1433?

oli-obk replied:
Whoops, yea, that's actually where I noticed the slice function is not sufficient

scottmcm commented on text/0000-is-aligned-intrinsic.md:
```diff
@@ -0,0 +1,287 @@
+## Example 2 (slices)
+
+The `memchr` impl in the standard library explicitly uses the three phases of
+the `align_to` functions:
```
bikeshed/typo: `align_to` here, `alignto` in other places.

CryZe replied:
Shouldn't it be called `align_to` in general? alignto feels unidiomatic.

oli-obk replied:
I kept mistyping it as `align_to`, too xD.

scottmcm commented on text/0000-is-aligned-intrinsic.md:
```diff
@@ -0,0 +1,287 @@
+#[repr(align = 64)]
+struct TwoWord([usize; 2]);
+
+let (head, mid, tail) = std::mem::align_to<TwoWord>(text);
```
Is the extra struct here necessary?  I'd have thought that just `align_to::<[usize;2], _>(text)` would be fine, since the original code only uses `usize`-alignment.

oli-obk replied:
Indeed, I misread that code. It aligns to usize, but steps in two usize chunks.

oli-obk commented on text/0000-is-aligned-intrinsic.md:
````diff
@@ -0,0 +1,287 @@
+Add a new intrinsic
+
+```rust
+fn alignto(ptr: *const (), align: usize) -> usize;
````
It's an intrinsic, so I'd guess invalid alignments are UB. I'd assume a non-power of two alignment or an alignment bigger than `2^32` would be considered invalid.

If that is ok, I'll add it as documentation. That's the rules that the compiler internal code has around alignment.

The alternative would be to add a generic argument for the pointer, but that would be confusing, since you have to cast the pointer to the target type, and the pointee type would change behaviour. Pointee types are easy to get wrong in casts.

oli-obk commented on text/0000-is-aligned-intrinsic.md:
```diff
@@ -0,0 +1,287 @@
+cannot be aligned. Since the caller needs to check whether the returned offset
+would be in-bounds of the allocation that the pointer points into, returning
+`usize::max_value()` will never be in-bounds of the allocation and therefor
+the caller cannot act upon the returned offset.
```
Yea, I considered this, but found that documenting that the result always needs to be checked much more realistic, especially since you might not want matches or `map` + closures in optimization critical code.

oli-obk commented on text/0000-is-aligned-intrinsic.md:
````diff
@@ -0,0 +1,287 @@
+
+```rust
+fn alignto(ptr: *const (), align: usize) -> usize {
+    align - (ptr as usize % align)
````
whoops... no.

oli-obk commented:
Bikeshed: Upvote this comment if you want `align_to`, downvote if you want to stay with `alignto`.

nagisa commented:
One way to ensure only powers of two are allowed in the align_to instruction is to take the power, not the alignment itself.

oli-obk commented:
@nagisa that would make it riddiculous to use... because you'd have to figure out the power of the result of `align_of`

est31 commented:
>  you'd have to figure out the power of the result of align_of

Easy:
```rust
0usize.leading_zeros() - (std::mem::size_of::<T>()).leading_zeros()
```

scottmcm commented:
Retagging as libs help would be good on the method to expose in core, but doesn't really affect lang.

You could also consider a T-compiler issue for just the intrinsic to start, and an RFC for the public API later.

scottmcm commented:
For the align parameter on the intrinsic, maybe it should just have the same rules as an align in Layout from Allocator? https://github.com/rust-lang/rfcs/blob/master/text/1398-kinds-of-allocators.md#layout-api

Using the power is neat, since aligning can be done with `lshr`+`shl nuw`, but I think it's too inconsistent with how alignments are represented elsewhere in Rust and in LLVM.

eddyb commented:
Why would shifts be faster than `let mask = align - 1; (x + mask) & mask`?

eternaleye commented:
@est31:

> 0usize.leading_zeros()

Oh dear 😛 

> Built-in Function: int __builtin_clz (unsigned int x)
>
> Returns the number of leading 0-bits in x, starting at the most significant bit position. If x is 0, the result is undefined.

LLVM doesn't say differently anywhere, either. Of course, that's just the intrinsic - Rust wraps it, I'd expect with a test for zero and a branch to `mem::size_of<T>()`.

Might as well save LLVM the trouble of optimizing the test out. 😛

est31 commented:
@eternaleye llvm allows you in fact to decide whether to have deterministic output or not: https://llvm.org/docs/LangRef.html#llvm-ctlz-intrinsic

I do not know whether that gets translated with a test or not, but I hope that LLVM is able to eval that whole expression at compilation time any way.

Just wanted to say that its not hard to figure out the power of a number for the base of two.

scottmcm commented:
@eddyb Probably they're all the same because it'll always be a (post-inlining) constant anyway.  Shifting would only be faster if the power-of-two-ness wasn't guaranteed, and thus the mod-to-mask conversion wasn't allowed.

@est31 Depends on the architecture.  i686 it's codegened as a test, but haswell has an instruction that means it doesn't need it.  (The test is output quite late, leading to occasionally-bad ASM, thus PR https://github.com/rust-lang/rust/pull/42556 and the existence of [`std::intrinsics::cttz_nonzero`](https://doc.rust-lang.org/nightly/std/intrinsics/fn.cttz_nonzero.html).)

oli-obk commented:
Would it be possible to get a new shepherd assigned?

scottmcm commented:
@KodrAus Interested in shepherding this?

ubsan commented:
Why is `align_to` in `mem`, not `ptr`?

KodrAus commented:
@scottmcm sure I can help shepherd the library function.

@ubsan `mem` seems like the most natural home for `align_to` to me.

ubsan commented:
@KodrAus why? you're not aligning memory, you're aligning pointers...

scottmcm commented:
Could also, after https://github.com/rust-lang/rfcs/pull/1966, be a method on pointers.

KodrAus commented:
@ubsan Yeh it seems like it could comfortably live in either place, but since you're reinterpreting slices `mem` seemed more natural to me than `ptr`. Would you be surprised to find `align_to` in `mem`?

The idea is to add `align_offset` to pointers, right? `align_to` could be an inherent method on slice and save having to fill in an inferrable generic argument, but it seems too niche for that. So just a function in `mem` with appropriate methods on `*const T` and `*mut T` like the RFC currently proposes sounds like a good API to me.

KodrAus commented on text/0000-is-aligned-intrinsic.md:
```diff
@@ -0,0 +1,294 @@
+let ptr = v.as_ptr();
+let align = unsafe {
+    // the offset is safe, because `index` is guaranteed inbounds
+    ptr.offset(index).align_offset(usize_bytes);
```
nit: This shouldn't have a `;`

KodrAus commented on text/0000-is-aligned-intrinsic.md:
```diff
@@ -0,0 +1,294 @@
+let len = text.len();
+let ptr = text.as_ptr();
+
+let (head, mid, tail) = std::mem::align_to::<(usize, usize)>(text);
```
nit: This would have to be: `std::mem::align_to::<(usize, usize), _>(text);`

KodrAus commented on text/0000-is-aligned-intrinsic.md:
```diff
@@ -0,0 +1,294 @@
+
+## Documentation
+
+A lint could be implemented which detects hand-written alignment checks and
```
Is this a `clippy` lint you're envisaging @oli-obk?

oli-obk replied:
Yea, I think it fits better into clippy. Currently rustc has no similar lint.

ubsan commented:
@KodrAus I'd imagine, if it dealt with slices, being in `std::slice`. `std::mem` seems like kind of a dumping ground for random functions, which I'd rather it not be.

KodrAus commented:
> `std::mem` seems like kind of a dumping ground for random functions, which I'd rather it not be.

I think you've got a good point there, `mem` is just a bunch of fundamental utilities _"...for querying the size and alignment of types, initializing and manipulating memory."_. Based on the other functions in that module, I still think it aligns (hehe) best with `mem`, but I'm also not against a function in `slice`. I just wouldn't look there for it first.

oli-obk commented:
>  I'd imagine, if it dealt with slices, being in std::slice. std::mem seems like kind of a dumping ground for random functions, which I'd rather it not be.

I can get behind that reasoning. The RFC has been changed accordingly.

KodrAus commented:
If we are going to put `align_to` in `slice` then I think it's worth exploring taking the step of making it an inherent method, since it does have a pretty clear receiver. It would save an easily inferrable generic argument, so you could have `text.align_to::<(usize, usize)>()` instead of `slice::align_to::<(usize, usize), _>()`.

Any thoughts?

oli-obk commented:
> Any thoughts?

I'm fine with it, but there's no precedent, so I worry a little if we want magic unsafe inherent methods

KodrAus commented:
That's a good question. There are a few unsafe inherent methods scattered around slice types, plenty on pointers since there's not much you can do with them that isn't unsafe. So I think there is actually a bit of precedent for inherent unsafe methods with unchecked indexing and slicing and whatnot.

From an ergonomics perspective it seems like a win to me.

scottmcm commented:
Aspirationally, It'd be great to have a _safe_ function that could do the `&[T] => (&[T], &[Chunk<T>], &[T])` split (maybe with an `N` on Chunk?).  That's probably impractical now, but it might affect the bikeshed.

oli-obk commented:
@scottmcm I think the only way to get that to be safe would be to have a runtime check for matching sizes and some form of `#[transparent]`, but it should be a trivial addition once this RFC is through and `#[transparent]` exists.

KodrAus commented:
Hmm, typically unsafe variants would have a prefix/suffix and the safe variant wouldn't, but `align_to_unsafe_mut` is a pretty ugly colour for a bikeshed... Assuming you would want both the safe and unsafe variants in that case.

I'm tempted to ignore a potential safe variant as far as this RFC is concerned and we can figure it out in the future when all the pieces are available to make such a thing possible.

Does that seem reasonable, or would you rather have some degree of future-proofing built in here?

oli-obk commented:
> Does that seem reasonable, or would you rather have some degree of future-proofing built in here?

I don't think we can create a safe variant without some form for unsafe marker traits and a canonical transparent aligning wrapper type. I'd think that the actual alignment function would not take a generic argument anymore and would probably be on the wrapper type instead of on slices. But in any case, I definitely do not think that it should be named `align_to`, since it'll do something slighty more complex than that.

KodrAus commented on text/0000-is-aligned-intrinsic.md:
```diff
@@ -0,0 +1,296 @@
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* produce a lint in case `sizeof<T>() % sizeof<U>() != 0` and in case the expansion
```
I think this would be another useful `clippy` lint to have. Otherwise you'll just end up with something like `(&[T], &[], &[])` which isn't useful, but also isn't the end of the world.

aturon commented:
@KodrAus, who has been shepherding this RFC, believes it's ready for consideration by the broader libs team. Therefore:

@rfcbot fcp merge

ubsan commented:
:+1:

scottmcm commented:
@KodrAus The `unsafe` version of this has all the same gotchas as `transmute_copy`, which is one of the most unsafe and discouraged functions in all of `std`.  And it might even have more traps because it can lead to non-byte-oriented reads that could be undef because of padding bytes.  I understand that it needs to behave the way it does, and it's certainly useful, so I have no objections to adding it, I just think it deserves a scarier name.  Maybe something along the lines of `aligned_transmute`?

KodrAus commented:
@scottmcm Personally I think the fact that the function is unsafe should be enough to make users and future readers go read the scarily worded docs it will have and understand those edges better. 

But at the same time, I think it's definitely worthwhile to look at ways to make the function communicate its edges better, especially given it'll live on a very commonly used type. If everyone is on-board with communicating the fact that this is transmuting in the method name itself then we can do another round of thinking about the name.

ubsan commented:
@scottmcm imo, the issue with `transmute` is not what it does - there are many ways to get `transmute`. The issue is with how *easy* it is. This is not easy, and people don't turn to it just because it's the easiest solution at the time to get their programs typechecking.

aturon commented:
This RFC has been **merged**! [Tracking issue](https://github.com/rust-lang/rust/issues/44488).
